index.js

```diff
@@ -160,7 +160,7 @@ const LIST_ITEM_END_R = / *\n+$/;
 const LIST_LOOKBEHIND_R = /(?:^|\n)( *)$/;
 const CAPTURE_LETTER_AFTER_HYPHEN = /-([a-z])?/gi;
 const NP_TABLE_R = /^(.*\|?.*)\n *(\|? *[-:]+ *\|[-| :]*)\n((?:.*\|.*\n)*)\n?/;
-const PARAGRAPH_R = /^((?:[^\n]|\n(?! *\n))+)(?:\n *)+\n/;
+const PARAGRAPH_R = /^((?:[^\n]+\n)+?(?=\n|\s*(?:[-*+]|\d{1,9}[.)])\s|#|>\s|[`~]{3,}(?=[^`~]*[`~]{3,})))(?: *\n+|\n*)/;
```

Comment:
There must be a simpler way to do this than jamming in all the other syntax detection regex?

Reply:
It's been a while since I worked on this, but let me see if I can remember.

A paragraph can span multiple lines, so we need to know when to terminate or it will just swallow up other text that should have been its own elements. 2 new lines is an easy terminator, but something about the order of execution meant that a paragraph followed by a header or list with only a single `\n` between them would consider those as part of the paragraph block, because it didn't know any better.

It's possible there could be some reordering or separating out the logic to identify an "upcoming" block element, but the way the code was modeled made it seem like the regex should match the entire paragraph element (and nothing more). To really know when to terminate in the current order of execution, it has to include all possible block identifiers, because we can't count on those blocks being picked up previously.